### DISCUSSION TITLE
currency() callback support

### DISCUSSION THREAD

**Discussion by chimit** [laravel/nova-issues#4645] [Q&A] [answered]
- Laravel Version: 9.21.6
- Nova Version: 4.12.12
- PHP Version: 8.1.7
- Operating System and Version: macOS Monterey 12.4
- Browser type and version: Safari macOS

### Description:

I can't specify the currency that is taken from a relationship:

```php
Currency::make('Monthly price', function () {
    return $this->prices()->monthly()->first()->amount;
})
    ->currency($this->prices()->monthly()->first()->currency),
```

Possible solution is to make `currency()` method accept callbacks. Currently, it only supports strings. Or, please, let me know if there are other ways to get it.

**Comment by crynobone** [comment, marked as answer]
You can try:

```php
Currency::make('Monthly Price')
    ->tap(function ($field) {
        $price = $this->prices()->monthly()->first();

        $field->resolveUsing(function () use ($price) {
            return $price->amount;
        })->currency($price->currency);
    })->exceptOnForms(),
```

**Reply by chimit** [reply]
Attempt to read property "currency" on null.

**Reply by crynobone** [reply]
You have to handle when `$price` is `null`.

**Reply by chimit** [reply]
But actually, it's not null.

**Reply by crynobone** [reply]
Your error message prove otherwise. In certain use-case a resource can be called via empty model, such as fetching fields for create view etc.

**Reply by chimit** [reply]
If I do something like this:

```php
Currency::make('Monthly price')
    ->tap(function ($field) {
        $price = $this->prices()->monthly()->first();

        if (!$price) {
            return;
        }

        info($price);

        $field->resolveUsing(function () use ($price) {
            return $price->amount;
        })->currency($price->currency);
    })->exceptOnForms(),
```

I don't get the error, but I still get an empty Monthly price field and see 5 invocations (correct) in logs.

**Reply by crynobone** [reply]
```php
Currency::make('Monthly price')
    ->tap(function ($field) {
        $price = $this->prices()->monthly()->first();

        $field->resolveUsing(function () use ($price) {
            return optional($price)->amount;
        })
        
        if (!is_null($price)) {
            $field->displayUsing(function () use ($field, $price) {
                return $field->formatMoney($price->amount);
            })->currency($price->currency);
        }
    })->exceptOnForms(),

**Reply by chimit** [reply]
Thanks! It works for the view page, but not for the index.

I think making `currency()` accept closures would still be a cleaner way (or somehow pass the currency from `make()`). So far I would instead use this simple approach:

```php
Text::make('Monthly price', function () {
    $price = $this->prices()->monthly()->first();

    return $price
        ? $price->currency . ' ' . number_format($price->amount)
        : null;
}),
```

**Reply by juse-less** [reply]
Pre-post: Darnit. It was resolved as I wrote below. >.<

I've had a lot of issues with this in the past and had to fallback to using a Number field and not display currency.
Now that I tried with the info that @crynobone gave above, it went smooth.
I've tried similar things before, so not sure what I did wrong (albeit, I wasn't using `tap()`).

In my case, though, all I needed to do was to check for is if the value/relation in the chain of calls was null.
I have an invoice that the currency is on, and the prices per line item where I want to display it.
It was failing on index views, but works with the snippet below (could've sworn I tried it in the past) -
```php
// Invoice Line Item
Currency::make(__('Unit Price'), attribute: 'unit_price')
    ->currency($this->invoice?->currency)
    ->sortable(),

Currency::make(__('Amount'), attribute: 'amount')
    ->currency($this->invoice?->currency)
    ->sortable(),
```

So you should be able to use something like the snippets from above.

**Reply by chimit** [reply]
What the... it works 😄

```php
Currency::make('Monthly price', function () {
    return $this->prices()->monthly()->first()?->amount;
})
    ->currency($this->prices()->monthly()->first()?->currency),
```

**Reply by juse-less** [reply]
I'd recommend calling it above the array, so you don't make duplicate database requests, though.
```php
public function fields()
{
    $monthlyPrice = $this->prices()->monthly()->first();

    return [
        Currency::make('Monthly price', fn () => $monthlyPrice?->amount)
            ->currency($monthlyPrice?->currency),
    ];
}
```
